internal/resolution/variable_sources/entity/bundle_entity.go

```diff
@@ -0,0 +1,217 @@
+	property.Channel
+	Replaces  string   `json:"replaces,omitempty"`
+	Skips     []string `json:"skips,omitempty"`
+	SkipRange string   `json:"skipRange,omitempty"`
+}
+
+type PackageRequired struct {
+	property.PackageRequired
+	SemverRange *semver.Range `json:"-"`
+}
+
+type GVK property.GVK
+
+func (g GVK) String() string {
+	return fmt.Sprintf(`group:"%s" version:"%s" kind:"%s"`, g.Group, g.Version, g.Kind)
+}
+
+type GVKRequired property.GVKRequired
+
+func (g GVKRequired) String() string {
+	return fmt.Sprintf(`group:"%s" version:"%s" kind:"%s"`, g.Group, g.Version, g.Kind)
+}
+
+type BundleEntity struct {
+	*input.Entity
+
+	// these properties are lazy loaded as they are requested
+	bundlePackage     *property.Package
+	providedGVKs      []GVK
+	requiredGVKs      []GVKRequired
+	requiredPackages  []PackageRequired
+	channelProperties *ChannelProperties
+	semVersion        *semver.Version
+	bundlePath        string
+	mu                sync.RWMutex
+}
+
+func NewBundleEntity(entity *input.Entity) *BundleEntity {
+	return &BundleEntity{
+		Entity: entity,
+		mu:     sync.RWMutex{},
+	}
+}
+
+func (b *BundleEntity) PackageName() (string, error) {
+	if err := b.loadPackage(); err != nil {
+		return "", err
+	}
+	return b.bundlePackage.PackageName, nil
+}
+
+func (b *BundleEntity) Version() (*semver.Version, error) {
+	if err := b.loadPackage(); err != nil {
+		return nil, err
+	}
+	return b.semVersion, nil
+}
+
+func (b *BundleEntity) ProvidedGVKs() ([]GVK, error) {
+	if err := b.loadProvidedGVKs(); err != nil {
+		return nil, err
+	}
+	return b.providedGVKs, nil
+}
+
+func (b *BundleEntity) RequiredGVKs() ([]GVKRequired, error) {
+	if err := b.loadRequiredGVKs(); err != nil {
+		return nil, err
+	}
+	return b.requiredGVKs, nil
+}
+
+func (b *BundleEntity) RequiredPackages() ([]PackageRequired, error) {
+	if err := b.loadRequiredPackages(); err != nil {
+		return nil, err
+	}
+	return b.requiredPackages, nil
+}
+
+func (b *BundleEntity) ChannelName() (string, error) {
+	if err := b.loadChannelProperties(); err != nil {
+		return "", err
+	}
+	return b.channelProperties.ChannelName, nil
+}
+
+func (b *BundleEntity) ChannelProperties() (*ChannelProperties, error) {
+	if err := b.loadChannelProperties(); err != nil {
+		return nil, err
+	}
+	return b.channelProperties, nil
+}
+
+func (b *BundleEntity) BundlePath() (string, error) {
+	if err := b.loadBundlePath(); err != nil {
+		return "", err
+	}
+	return b.bundlePath, nil
+}
+
+func (b *BundleEntity) loadPackage() error {
+	if b.bundlePackage == nil {
+		bundlePackage, err := loadFromEntity[property.Package](b.Entity, property.TypePackage)
+		if err != nil {
+			return fmt.Errorf("error determining package for entity '%s': %s", b.ID, err)
+		}
+		b.bundlePackage = &bundlePackage
+		if b.semVersion == nil {
+			semVer, err := semver.Parse(b.bundlePackage.Version)
+			if err != nil {
+				return fmt.Errorf("could not parse semver (%s) for entity '%s': %s", b.bundlePackage.Version, b.ID, err)
```

Comment:
Nit: use `%w` when wrapping `error`s in `fmt.Errorf()`.
Here and below.

Reply:
Good call! Fixed =D

Reply:
In case anyone is wondering why you should use `%w` when wrapping errors in `fmt.Errorf()`, checkout this [blog](https://gosamples.dev/wrap-unwrap-errors/).